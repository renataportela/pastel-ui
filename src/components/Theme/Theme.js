import React from 'react'
import PropTypes from 'prop-types'
import { createGlobalStyle, css } from 'styled-components'

import defaultTheme from '~/styles/theme'

export const ThemeContext = React.createContext()

function Theme({ children, theme }) {
  const appTheme = theme ? {...defaultTheme, ...theme} : defaultTheme;

  return (
    <ThemeContext.Provider value={{ ...appTheme }}>
      <GlobalStyle theme={appTheme} />
      {children}
    </ThemeContext.Provider>
  )
}

const GlobalStyle = createGlobalStyle`
  *,
  *::before,
  *::after {
    box-sizing: border-box;
  }

  html {
    scroll-behavior: smooth;
  }

  body {
    margin: 0;
    padding: 0;
    font-weight: 400;
    line-height: 1.5;
    ${({ theme }) => css`
      font-family: ${theme.fontFamily};
      font-size: ${theme.fontSize}px;
      color: ${theme.colors.text};
      background-color: ${theme.colors.bgColor};
    `}
  }

  [tabindex='-1']:focus:not(:focus-visible) {
    outline: 0 !important;
  }
`

Theme.propTypes = {
  children: PropTypes.node.isRequired,
  theme: PropTypes.shape({
    fontFamily: PropTypes.string,
    fontSize: PropTypes.number,
    colors: PropTypes.object,
  }),
}

export default Theme
