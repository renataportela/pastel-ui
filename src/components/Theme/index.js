export { default } from './Theme';
export { default as BaseColors, BASE_COLORS_PROPS } from './BaseColors';
export { default as useTheme } from './useTheme';