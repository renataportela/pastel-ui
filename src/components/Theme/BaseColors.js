import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { PALLETE } from '~/constants'
import useTheme from './useTheme'

function BaseColors({ bgColor, color, textColor, ...props }) {
  const { colors } = useTheme()
  const palleteColor = getColor(colors, color) || { text: 'currentColor', bg: null };
  const fontColor = getColor(colors, textColor, 'text') || textColor;
  const bg = getColor(colors, bgColor, 'bg') || bgColor;
  
  return (
    <BaseStyle 
      $textColor={fontColor || palleteColor.text} 
      $bgColor={bg || palleteColor.bg} 
      {...props} 
    />
  )
}

const getColor = (colors, chosen, attr) => {
  if (!chosen) return null
  const color = colors[chosen];
  return color && attr ? color[attr] : color;
}

const BaseStyle = styled.div(({ $bgColor, $textColor }) => {
  return {
    backgroundColor: $bgColor,
    color: $textColor,
  }
})

export const BASE_COLORS_PROPS = {
  bgColor: PropTypes.string,
  color: PropTypes.oneOf(PALLETE),
  textColor: PropTypes.string,
}

BaseColors.propTypes = BASE_COLORS_PROPS

export default BaseColors