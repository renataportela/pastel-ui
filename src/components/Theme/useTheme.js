import { useContext } from 'react'

import { ThemeContext } from './Theme'

function useTheme() {
  return useContext(ThemeContext)
}

export default useTheme