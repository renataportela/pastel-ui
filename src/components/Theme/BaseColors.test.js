import React from 'react'
import renderer from 'react-test-renderer'

import Theme from './Theme'
import BaseColors from './BaseColors'

describe('Div with background and text colors', () => {
  it('renders colors by pallete name', () => {
    const tree = renderer
      .create(<Theme><BaseColors color="warning" /></Theme>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders colors by pallete name separately', () => {
    const tree = renderer
      .create(<Theme><BaseColors bgColor="primary" textColor="warning" /></Theme>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders colors by hexadecimal number', () => {
    const tree = renderer
      .create(<Theme><BaseColors bgColor="#ad0" textColor="#000" /></Theme>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })  
})