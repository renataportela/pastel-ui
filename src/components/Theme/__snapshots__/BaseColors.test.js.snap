// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Div with background and text colors renders colors by hexadecimal number 1`] = `
<div
  className="BaseColors__BaseStyle-s1fbie-0 kNLDZe"
/>
`;

exports[`Div with background and text colors renders colors by pallete name 1`] = `
<div
  className="BaseColors__BaseStyle-s1fbie-0 jTDUIl"
/>
`;

exports[`Div with background and text colors renders colors by pallete name separately 1`] = `
<div
  className="BaseColors__BaseStyle-s1fbie-0 cKbdWp"
/>
`;
